components/proxy/src/main/java/com/hotels/styx/proxy/backends/file/FileChangeMonitor.java

```diff
@@ -0,0 +1,155 @@
+ * limitations under the License.
+ */
+package com.hotels.styx.proxy.backends.file;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.hash.HashCode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.attribute.FileTime;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.hotels.styx.common.Files.fileContentMd5;
+import static java.lang.String.format;
+import static java.nio.file.Files.exists;
+import static java.nio.file.Files.getLastModifiedTime;
+import static java.nio.file.Files.isReadable;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+/**
+ * Monitors a file system object and notifies the consumer of any changes.
+ */
+public class FileChangeMonitor implements FileMonitor {
+    private final Path monitoredFile;
+    private final ScheduledExecutorService executor = newSingleThreadScheduledExecutor();
+    private final long pollPeriod;
+    private final TimeUnit timeUnit;
+    private ScheduledFuture<?> monitoredTask;
+
+    private AtomicReference<FileTime> lastChangedTime = new AtomicReference<>(FileTime.fromMillis(0));
+    private AtomicReference<HashCode> hashCode = new AtomicReference<>();
+    private volatile boolean performHashCheck;
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FileChangeMonitor.class);
+
+    @VisibleForTesting
+    FileChangeMonitor(String monitoredFile, long pollPeriod, TimeUnit timeUnit) {
+        requireExists(requireNonNull(monitoredFile));
+        this.monitoredFile = Paths.get(monitoredFile);
+        this.pollPeriod = pollPeriod;
+        this.timeUnit = requireNonNull(timeUnit);
+        this.hashCode.set(HashCode.fromLong(0));
+    }
+
+    public FileChangeMonitor(String monitoredFile) {
+        this(monitoredFile, 1, SECONDS);
+    }
+
+    @Override
+    public void start(Listener listener) {
+        synchronized (this) {
+            if (monitoredTask != null) {
+                throw new IllegalStateException(format("File monitor for '{}' is already started", monitoredFile));
+            }
+
+            monitoredTask = executor.scheduleAtFixedRate(newPollerTask(listener), pollPeriod, pollPeriod, timeUnit);
+        }
+    }
+
+    private Runnable newPollerTask(Listener listener) {
+        return () -> {
+            if (!exists(monitoredFile)) {
+                LOGGER.debug("Monitored file does not exist. Path={}", monitoredFile);
+                return;
```

Comment:
Could we not just use `else`?

```
if(condition) {
  // code

} else if(otherCondition) {
  // more code

} else ...

Reply:
Matter of taste. IMO the function would look very crowded with  if/else branching.

Reply:
Well here's a comparison:

**Before:**
```
            if (!exists(monitoredFile)) {
                LOGGER.debug("Monitored file does not exist. Path={}", monitoredFile);
                return;
            }

            if (!isReadable(monitoredFile)) {
                LOGGER.debug("Monitored file is no longer readable. Path={}", monitoredFile);
                return;
            }

            if (modificationTimeChanged(monitoredFile)) {
                hashCode.set(fileContentMd5(monitoredFile));
                performHashCheck = true;

                listener.fileChanged();
                return;
            }

            if (performHashCheck && contentHashChanged(monitoredFile)) {
                listener.fileChanged();
            }
```
**After:**
```
            if (!exists(monitoredFile)) {
                LOGGER.debug("Monitored file does not exist. Path={}", monitoredFile);

            } else if (!isReadable(monitoredFile)) {
                LOGGER.debug("Monitored file is no longer readable. Path={}", monitoredFile);

            } else if (modificationTimeChanged(monitoredFile)) {
                hashCode.set(fileContentMd5(monitoredFile));
                performHashCheck = true;

                listener.fileChanged();

            } else if (performHashCheck && contentHashChanged(monitoredFile)) {
                listener.fileChanged();
            }
```

I don't think there's a huge difference, but the latter removes some unnecessary lines/statements.